### ISSUE TITLE
i3-wm does not work properly with some applications in full-screen mode

### ISSUE THREAD

**Issue by bonidydy** [i3/i3#2019]
i3-wm has a problem with some applications running in full-screen mode. To reproduce it please start some virtual machine (with guest additions) in VirtualBox and then enter full-screen mode (initiated from VirtualBox menu).  The screen gets blank.  Sometimes the whole workspace gets blocked.  The same happens with google-chrome when some pop-up manages to break through.  It is however more difficult to reproduce .  I am not 100% sure if the problem is caused by i3 only.  In the meantime there have been several updates to i3, virtualbox and chrome (Debian sid).  I can just say it does not happen with other window managers now and did not occur in earlier versions of i3 (4.10).

http://logs.i3wm.org/logs/5138033674289152.bz2

i3 version 4.11 (2015-09-30, branch "4.11") © 2009 Michael Stapelberg and contributors

**Comment by i3bot** [comment]
I don’t see a link to logs.i3wm.org. Did you follow http://i3wm.org/docs/debugging.html? (In case you actually provided a link to a logfile, please ignore me.)

**Comment by i3bot** [comment]
I don’t see a version number. Could you please copy & paste the output of `i3 --version` into this issue?

**Comment by Airblader** [comment]
Hoe does this differ from #1979 , #1946

**Comment by bonidydy** [comment]
Differences:
#1979 - toggling full screens does not help; I have to change workspace and kill chrome; I think it started to happen with chrome 46
#1946 - not using VirtualBox in seamless mode; it is just full screen mode with guest resized; screen gets blank not garbled or distorted; if workspace is not blocked and VirtualBox menu or shortcuts work I can switch back to windowed mode which is all right again

**Comment by Airblader** [comment]
I still feel like the issues are similar enough, but you actually mentioned something important:

> and did not occur in earlier versions of i3 (4.10).

Please do a git bisect to find out which commit broke this.

**Comment by bonidydy** [comment]
Sorry I am not a programmer. I'm afraid I won't be able to find it out even if I dig through the commits :-(.

**Comment by Airblader** [comment]
You don't need to dig through commits or understand the source code at all. Just read [this](https://robots.thoughtbot.com/git-bisect).

In order to build i3 locally, install git and clone it somewhere via `git clone https://github.com/i3/i3.git`. Build and install it via `make && sudo make install`. Then restart i3, see if the bug persists and follow the bisecting guide accordingly.

When you're done, simply reinstall i3 from your package manager to get back to the current 4.11 release.

**Comment by bonidydy** [comment]
I did not do that bisect thing but that gave me another idea. I downgraded i3 to 4.8 from a Debian snapshot. The behavior of chrome and virtualbox was still equally wrong. Then I downgraded virtualbox 5.0.6 to 4.3.30 and it started working properly. I cannot do the same with chrome but I am pretty sure it started with the newest version 46.

So it is not a change in i3 that breaks things. Nonetheless there is something wrong as other window managers work properly in described situations.

**Comment by Airblader** [comment]
Thanks. This sounds like the applications should take a look at it first. Please file bug reports there and link them here. 

I'm going to close this issue because the indication is a client side bug. Once we have proof or reason to believe i3 is doing something wrong, we can revisit this. It working in other window managers isn't a much better argument than other applications working fine in fullscreen.

On top of all that i3 is not involved in the rendering of the client whatsoever so it seems even less likely to be an i3 issue.

**Comment by jarhat** [comment]
For the virtualbox + i3-wm fullscreen issue there's a workaround. 

-> http://i.imgur.com/3CAmijr.png

I've found it here..

https://bbs.archlinux.org/viewtopic.php?id=200283

**Comment by Airblader** [comment]
Thanks for the information. This seems to confirm that it isn't i3's fault, seeing how people seem to have the same issue on other setups as well.

**Comment by melynx** [comment]
The chrome issue is still present though and there seems to be no way around it...
This might be yet another related bug. https://github.com/i3/i3/issues/1751

i3 version 4.10.4 (2015-09-08, branch "4.10.4") © 2009-2014 Michael Stapelberg and contributors

**Comment by ffernand** [comment]
The chrome issue is still present even for the latest version of i3.  Ctrl-n creates a new window not visible during chrome full screen mode.  Current full screen is frozen (ie. accepts no keyboard or mouse input) until full screen mode is existed.  Then both the original chrome window and the new window show up size by size.

This happens even for popups that chrome might generate in the course of browsing.  An example is Jira popups that provide application help but in full screen mode, the window is missing and the main window is unresponsive until full screen mode is exited.

The setting *popup_during_fullscreen* doesn't work for either *smart* or *leave_fullscreen* when using chrome.

$ i3 --version
i3 version 4.11 (2015-09-30, branch "4.11") © 2009 Michael Stapelberg and contributors

**Comment by Airblader** [comment]
We know the issue exists, but we currently don't think it's an i3 issue.

**Comment by ffernand** [comment]
> We know the issue exists, but we currently don't think it's an i3 issue.

I'm not sure if you mean that it's an issue with just Chrome or if it's an issue with apps in full screen in general.  A friend of mine did confirm that the Chrome issue in full screen exists in Xmonad as well, but I've noticed that it's not limited to Chrome either.

Another example is Gimp.  Opening the 'Preferences' dialog via 'Edit -> Preferences' or the 'Open' dialog via 'File -> New' when Gimp is in full screen mode doesn't show the dialog box, until I exit the full screen.  Neither the *smart* or *leave_fullscreen* settings seems to have an effect when applied against *popup_during_fullscreen*.

It's also the same with Inkscape in *smart* mode though Inkscape does seem to obey the *leave_fullscreen* setting.

In any case, if it's entirely the apps fault, full screen mode does present itself as rather unreliable when engaged in an activity that can potentially open a window.

I don't know what's involved but I imagine it's hard to tell if a new window belongs to the full screen app or not.  But is it feasible to consider having all new floating windows (even if it can't be identified as belonging to the current app) to bubble to the top in full screen mode?

**Comment by Airblader** [comment]
Generally: if it happens for different applications, a good way to start is looking what they have in common. For example, if they all use a toolkit like GTK, it might be a bug in GTK.

> I don't know what's involved but I imagine it's hard to tell if a new window belongs to the full screen app or not.

For things like dialogs, there are [WM_TRANSIENT_FOR](https://tronche.com/gui/x/icccm/sec-4.html#WM_TRANSIENT_FOR) and [WM_CLIENT_LEADER](https://tronche.com/gui/x/icccm/sec-5.html), which are also the things i3 looks at for the `popup_during_fullscreen` directive.

I've also looked a bit into the Gimp issue you mentioned now since it's so nice and easy to reproduce. Gimp sets the `WM_CLIENT_LEADER` property, but it sets it to an unmapped placeholder window. Since i3 doesn't know (care) about unmapped containers, the popup handling doesn't take care of it.

@stapelberg Actually I'm currently not entirely sure why for `popup_during_fullscreen leave_fullscreen` we demand that the leader window is mapped. Do you know if there's a specific reason for it? AFAICT we could toggle fullscreen in this case whether or not the leader is managed, as long as the leader is set (and not to itself).

**Comment by Airblader** [comment]
Just FYI, for gimp the following patch (which is pretty ugly) works:

````
diff --git a/src/manage.c b/src/manage.c
index cba395e..a691a6e 100644
--- a/src/manage.c
+++ b/src/manage.c
@@ -420,6 +420,8 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
                 transient_win = next_transient->window;
             }
         }
+    } else if (cwindow->leader != XCB_NONE && cwindow->leader != cwindow->id && config.popup_during_fullscreen == PDF_LEAVE_FULLSCREEN && fs != NULL) {
+        con_toggle_fullscreen(fs, CF_OUTPUT);
     }
 
     /* dock clients cannot be floating, that makes no sense */
```

For Chrome, however, it doesn't work, because it doesn't set a leader at all (which, frankly, makes sense). I think the issue with Chrome really lies within Chrome because it shouldn't block the donating window until the new window has received focus. That makes no sense to me and I'd consider that a bug in Chromium.

**Comment by KiddyTheKid** [comment]
I leave this as one more answer, sorry if it is a far from elegant solution. I had the same problem, installing on mint 21.1 straight from the repositories; what I did was to completely uninstall i3 after backing up my i3 config file, just in case: i3, which uninstalls i3-lock during autoremove; i3-wm, which uninstalls dex, libanyevent-i3-perl, libxcb-xrm0. 
I restarted the computer and reinstalled i3, restarted the computer again and started my user session with i3 and now I can use full screen mode without problems (for now).
